src/rawposix/src/fs_calls.rs

```diff
@@ -1554,345 +1585,538 @@ pub fn fdatasync_syscall(
     arg6: u64,
     arg6_cageid: u64,
 ) -> i32 {
-    // Type conversion
-    let virtual_fd = sc_convert_sysarg_to_i32(fd_arg, fd_cageid, cageid);
+    let kernel_fd = convert_fd_to_host(vfd_arg, vfd_cageid, cageid);
+    // Return error
+    if kernel_fd < 0 {
+        return handle_errno(kernel_fd, "fchdir");
+    }
 
-    // Validate unused args
     if !(sc_unusedarg(arg2, arg2_cageid)
         && sc_unusedarg(arg3, arg3_cageid)
         && sc_unusedarg(arg4, arg4_cageid)
         && sc_unusedarg(arg5, arg5_cageid)
         && sc_unusedarg(arg6, arg6_cageid))
     {
-        panic!("{}: unused arguments contain unexpected values -- security violation", "fdatasync_syscall");
-    }
-
-    let kernel_fd = convert_fd_to_host(virtual_fd as u64, fd_cageid, cageid);
-    // Return error 
-    if kernel_fd < 0 {
-        return handle_errno(kernel_fd, "read");
+        panic!("{}: unused arguments contain unexpected values -- security violation", "fchdir_syscall");
     }
 
-    let ret = unsafe { libc::fdatasync(kernel_fd) };
-
+    let ret = unsafe { libc::fchdir(kernel_fd) };
     if ret < 0 {
-        let errno = get_errno();
-        return handle_errno(errno, "fdatasync");
+        return handle_errno(get_errno(), "fchdir");
+    }
+
+    // Update the cage's current working directory
+    // We need to get the current working directory from the kernel to update the cage
+    let mut cwd_buf = [0u8; PATH_MAX as usize];
+    let cwd_ptr = unsafe { libc::getcwd(cwd_buf.as_mut_ptr() as *mut i8, cwd_buf.len()) };
+    if !cwd_ptr.is_null() {
+        if let Some(cage) = get_cage(cageid) {
+            let host_path = unsafe { std::ffi::CStr::from_ptr(cwd_ptr) }.to_string_lossy();
```

Comment:
It's weird to me that we need to use unsafe here, also seems like it should be a helper function